### PR TITLE
Introducing React Hooks for AppInsights

### PR DESCRIPTION
**Pull request by aaronpowell**
_Targeting issue #991._

This PR introduces some new features to the React (DOM) package, custom hooks that can be used to track an event or metrics on a component (metric tracking is similar to the HOC `withAITracking`).

It is based off the [blog post I wrote](https://dev.to/aaronpowell/combining-react-hooks-with-appinsights-5692) on the subject of Hooks + AppInsights.

### CODE REVIEW

**Comment by aaronpowell** [comment]
Presently this PR will be blocked as the version of TypeScript in use, 2.5.3, does not support [Conditional Types](http://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types) (they were added in 2.8).

This was uncovered as I had to update the `@types/react` packages in use because it was using type definitions from pre-hooks React, but the type definitions use Conditional Types.

File: extensions/applicationinsights-react-js/src/useTrackEvent.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * @copyright Microsoft 2019
+ */
+import { useState, useEffect, useRef } from "react";
+import ReactPlugin from "./ReactPlugin";
+
+export default function useCustomEvent<T>(
+  reactPlugin: ReactPlugin,
+  eventName: string,
+  eventData: T,
+  skipFirstRun = true
```
Comment (xiao-lix) [inline extensions/applicationinsights-react-js/src/useTrackEvent.ts]: @aaronpowell do you mind explaining why this `skipFirstRun` flag is needed here? thanks.

Comment (aaronpowell) [reply extensions/applicationinsights-react-js/src/useTrackEvent.ts]: @xiao-lix The reason is to mimic more closely the way the non-hook version works. With `useEffect` hooks the effect is triggered on each value update _including_ the initial setting of the value, meaning we would fire off too early and potentially have unwanted events tracked. I explain it more in the post I wrote: https://www.aaron-powell.com/posts/2019-11-19-combining-react-hooks-with-appinsights/#ignoring-unwanted-effect-runs

**Comment by bschlintz** [comment]
Is there any update on this PR @xiao-lix? I have a need for this functionality in a project I'm working on with a customer.

File: extensions/applicationinsights-react-js/src/ReactPlugin.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ export default class ReactPlugin extends BaseTelemetryPlugin {
 
     private _analyticsPlugin: IAppInsights;
     private _extensionConfig: IReactExtensionConfig;
+    private _logger: IDiagnosticLogger;
```
Comment (MSNev) [inline extensions/applicationinsights-react-js/src/ReactPlugin.ts:23]: You don't need to do this.

It's already exposed as
this.diagLog()

File: extensions/applicationinsights-react-js/src/ReactPlugin.ts [hunk 3]
```diff
@@ -69,6 +71,15 @@ export default class ReactPlugin extends BaseTelemetryPlugin {
         }
     }
 
+    trackEvent(event: IEventTelemetry, customProperties?: ICustomProperties) {
+        if (this._analyticsPlugin) {
+            this._analyticsPlugin.trackEvent(event, customProperties);
+        } else {
+            this._logger.throwInternal(
```
Comment (MSNev) [inline extensions/applicationinsights-react-js/src/ReactPlugin.ts:78]: So this then becomes

this.diagLog().throwInternal(...)

Comment (MSNev) [reply extensions/applicationinsights-react-js/src/ReactPlugin.ts:78]: I'm not seeing any other blocking issues apart from the intro of the _logger.
For references the diaLog() takes care of the this getting called before initialization() as well.

**Pull request merged**